### PR TITLE
load words dynamically

### PR DESCRIPTION
**Pull request by omarudolley**
Load word file based on selected language

### CODE REVIEW

File: frontend/src/lib/stores.js [hunk 1]
```diff
@@ -1,20 +1,38 @@
...
+  return (await import(`../languages/${code}/words.json`)).default
+}
 
 function randInt(min, max) {
   return Math.floor(Math.random() * (max - min)) + min
 }
 
-function getRandomWord() {
+async function getRandomWord() {
+  let words = await loadWords(_currentLanguage.code)
```
Comment (lietu) [inline frontend/src/lib/stores.js:17]: `loadWords` should only be called when the language changes, i.e. in `currentLanguage.subscribe()`

Comment (lietu) [reply frontend/src/lib/stores.js:17]: Or rather, the `import` should only be called when the language has changed, so you could also implement the caching for that in that function

Comment (omarudolley) [reply frontend/src/lib/stores.js]: if `loadWords` should only be called when the language changes, how do we initialize `words `?

File: frontend/src/lib/stores.js [hunk 2]
```diff
@@ -1,27 +1,41 @@
 import { writable } from 'svelte/store'
 import languages from '../languages/languages.json'
 
-export const currentLanguage = writable(languages[0])
-let _currentLanguage
-currentLanguage.subscribe((value) => (_currentLanguage = value))
+let words = [
```
Comment (omarudolley) [inline frontend/src/lib/stores.js:4]: This doesn't look great but it breaks if we don't initialize words.

File: frontend/src/lib/stores.js [hunk 3]
```diff
@@ -1,5 +1,28 @@
...
+      {
+        word: 'Hi',
+        tag: 'NOUN',
+      },
+    ],
+  },
+]
+
+async function loadWords() {
+  return (await import(`../languages/${_currentLanguage.code}/words.json`)).default
```
Comment (lietu) [inline frontend/src/lib/stores.js:17]: Make this function check if the currently loaded words are already for the correct language, if they are, then return them. If they are not, then import them and then return them.

File: frontend/src/lib/stores.js (hunk 2 above)
Comment (lietu) [reply frontend/src/lib/stores.js:4]: Yeah not like this.

You'll need something like

```js
let wordCache
let cachedLanguage
```

and then

```js
async function loadWords(language) {
  if (cachedLanguage !== language)  {
    wordCache = await ...
    cachedLanguage = language
  }
  return wordCache
}
```

Comment (lietu) [reply frontend/src/lib/stores.js:4]: though I'd change the "loadWords" name to "getWords"

Comment (lietu) [reply frontend/src/lib/stores.js:4]: oh yeah, you don't need `language` argument as you can access `_currentLanguage`

File: frontend/src/lib/stores.js [hunk 4]
```diff
@@ -1,12 +1,31 @@
...
+  await getWords()
+})
 
 function randInt(min, max) {
   return Math.floor(Math.random() * (max - min)) + min
 }
 
 function getRandomWord() {
-  return words[randInt(0, words.length - 1)]
+  return wordCache[randInt(0, wordCache.length - 1)]
```
Comment (omarudolley) [inline frontend/src/lib/stores.js:28]: I am probably missing something. The app breaks at this line because `wordCache` is `undefined`.

Comment (omarudolley) [reply frontend/src/lib/stores.js:28]: I think we have to initialize `wordCache`

Comment (lietu) [reply frontend/src/lib/stores.js:28]: I didn't say you should directly access `wordCache`, but `const words = await getWords()` and `getWords` has the caching logic inside it

Comment (omarudolley) [reply frontend/src/lib/stores.js:28]: where am I going to call this `const words = await getWords()`? in the `getRandomWord `function?_ it has to be in an async function.

Comment (lietu) [reply frontend/src/lib/stores.js:28]: yes

Comment (omarudolley) [reply frontend/src/lib/stores.js]: okay resolved.

File: frontend/src/lib/stores.js [hunk 5]
```diff
@@ -1,20 +1,49 @@
-import words from '../languages/et/words.json'
 import { writable } from 'svelte/store'
+import languages from '../languages/languages.json'
+
+let wordCache
+let cachedLanguage
+
+async function getWords() {
```
Comment (lietu) [inline frontend/src/lib/stores.js:7]: Just one last thing, I'd move this function to be above `randInt` so the `_currentLanguage` stuff is defined above it and it isn't as unclear

**Review by lietu — APPROVED** [review]

File: frontend/src/lib/stores.js (hunk 5 above)
Comment (omarudolley) [reply frontend/src/lib/stores.js:14]: ok

**Pull request merged**